Convert 1x1 <img> to <amp-pixel>

FuzzyTree opened the pull request:
This PR converts 1x1 images to [amp-pixel](https://www.ampproject.org/docs/reference/amp-pixel.html) tags (instead of amp-img).

FuzzyTree committed:
convert 1x1 img to amp-pixel

sebastianbenz commented on tests/test-data/fragment-html/img-test-fragment.html.out:
```diff
@@ -21,6 +21,9 @@
 <amp-img layout="responsive" src="https://upload.wikimedia.org/wikipedia/commons/e/ee/%22Birdcatcher%22_with_jockey_up.jpg" width="625" height="480"></amp-img>
 
+<!-- should transform to amp-pixel -->
+<amp-pixel src="https://upload.wikimedia.org/wikipedia/commons/e/ee/%22Birdcatcher%22_with_jockey_up.jpg" width="1" height="1"></amp-pixel>
```
you don't need to specify width and height.

sebastianbenz commented on src/Pass/ImgTagTransformPass.php:
```diff
@@ -72,22 +72,31 @@ function pass()
-                // Abort the conversion and remove the new img tag
-                $new_el->remove();
-                continue;
+            if ($el->attr('width') === '1' && $el->attr('height') === '1') {
```
extract `$el->attr('width') === '1' && $el->attr('height') === '1'` into a method named: `isPixel`.

sebastianbenz reviewed (commented):
Few minor suggestions - otherwise looks good!

FuzzyTree committed:
Replace image for amp-pixel test with actual 1x1 image
Remove width & height attributes from img tag in amp-pixel unit test
Add isPixel function to ImgTagTransformPass

FuzzyTree committed:
isPixel: return false if size can not be determined

FuzzyTree commented:
@sebastianbenz Thanks for reviewing. I've created `isPixel` and have removed `width` and `height` attributes. Removing those attributes meant I had to add a call to `setResponsiveImgHeightAndWidth` inside of `isPixel`.

Let me know if any other changes are needed.

sebastianbenz commented on src/Pass/ImgTagTransformPass.php:
```diff
@@ -72,22 +72,33 @@ function pass()
-                // Abort the conversion and remove the new img tag
-                $new_el->remove();
-                continue;
+            if ($this->isPixel($el)) {
```
Seems like github swallowed my other comment. Here is it again: can you please extract each if branch into a separate method. e.g.

```
if ($this->isPixel($el)) {
   convertAmpPixel(...)
} else {
  convertAmpImg(...)
} 
```

This makes it a bit easier to understand

sebastianbenz commented on src/Pass/ImgTagTransformPass.php:
```diff
@@ -133,6 +144,19 @@ protected function isSvg(\DOMElement $el)
+     */
+    protected function isPixel(DOMQuery $el)
+    {
+        if (!$this->setResponsiveImgHeightAndWidth($el)) {
```
I'm not sure this works in all cases. I think it's cleaner if you remove the width and height attribute from the converted element.

FuzzyTree replied:
Are you suggesting to remove the height and width attributes if they were not already set so that `isPixel` doesn't modify attributes of `$el`? Sorry, I don't completely understand.

sebastianbenz replied:
amp-pixel doesn't require width and height -> it's safe to remove these attributes.

FuzzyTree replied:
The generated amp-pixel tag will have `src` as its only attribute because instead of cloning `$el` with `cloneAndRenameDomElement` a new amp-pixel tag is created:

```
$new_dom_el = $dom_el->ownerDocument->createElement('amp-pixel');
$new_dom_el->setAttribute('src', $el->attr('src'));
$dom_el->parentNode->insertBefore($new_dom_el, $dom_el);
```

sebastianbenz replied:
My fault - should've re-read your code. The problem lies somewhere else: you duplicate the calculation of the image dimensions in convertAmpImg and convertAmpPixel. You could avoid this by moving this at the beginning instead:

```
$hasWidthAndHeight = $this->setResponsiveImgHeightAndWidth($new_el);
if (!$hasWidthAndHeight) {
  $this->addActionTaken(new ActionTakenLine('img', ActionTakenType::IMG_COULD_NOT_BE_CONVERTED, $lineno, $context_string));
  return false
}
if (isPixel(...)) {
  ...
}
```

FuzzyTree committed:
add convertAmpImg and convertAmpPixel functions

FuzzyTree committed:
Move setResponsiveImgHeightAndWidth to top so it's not called twice

sebastianbenz commented on src/Pass/ImgTagTransformPass.php:
```diff
@@ -66,33 +66,49 @@ function pass()
-
+            if ($this->isPixel($el)) {
+                $new_dom_el = $this->convertAmpPixel($el);
+                $this->context->addLineAssociation($new_dom_el, $lineno);
```
remove `$this->context->addLineAssociation($new_dom_el, $lineno);` from the if branches and do it once before `$el->remove();`

sebastianbenz commented on src/Pass/ImgTagTransformPass.php:
```diff
@@ -66,33 +66,49 @@ function pass()
+            if ($this->isPixel($el)) {
+                $new_dom_el = $this->convertAmpPixel($el);
+                $this->context->addLineAssociation($new_dom_el, $lineno);
+                $this->addActionTaken(new ActionTakenLine('img', ActionTakenType::IMG_PIXEL_CONVERTED, $lineno, $context_string));
```
move inside convertAmpPixel

sebastianbenz commented on src/Pass/ImgTagTransformPass.php:
```diff
@@ -66,33 +66,49 @@ function pass()
+            else {
+                $new_dom_el = $this->convertAmpImg($el);
+                $this->context->addLineAssociation($new_dom_el, $lineno);
+                $this->addActionTaken(new ActionTakenLine('img', ActionTakenType::IMG_CONVERTED, $lineno, $context_string));
```
move inside convertAmpImg

FuzzyTree committed:
Move addActionTaken lines inside convertAmp* functions

FuzzyTree committed:
Move addLineAssociation lines to before $el->remove() call

FuzzyTree committed:
Fix unit test. Remove unnecessary call from isPixel

FuzzyTree committed:
$hasWidthAndHeight => $has_height_and_width for consistency in code style

FuzzyTree committed:
code style: opening function { on new line

FuzzyTree committed:
Add function comments

sebastianbenz commented:
Thanks!
